# Main Work Flow
Keep all the collaborated work here

## Data Retrieval
SRA files from NCBI for each set of samples were downloaded and converted to fastq.gz format. See [notes](https://github.com/srmarzec/albopictus_remapping/blob/main/misc/sra_accession/sraRetrievalTips.md) for details.

## Cleaning Data
Preliminary tests showed no need for preprocessing raw data beside trimmomatic steps (read alignment rate is ~82%).

### Trimmomatic
Trimmomatic version 0.39 [code for download](https://github.com/srmarzec/albopictus_remapping/blob/main/misc/SarahNotes.md#downloading-the-right-version-of-trimmomatic)

Generic script: [test_trim.sh](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/test_trim.sh)

We decided to use this adapter file because based on the manual, it was most likely used for HiSeq machines (which all sets of our data were run on). Note that all parameter for trimmomatic will run in the order they are written.

### Quality Control

Fastqc

## Mapping
We will be mapping with STAR (v2.7.1a)
Our reference genome is a new one generated by PI's group but we have access through him.
With Star, the genome should be indexed first: [STAR_genomeIndex.sh](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/STAR_genomeIndex.sh)
Then mapped:

