# Main Work Flow
Keep all the collaborated work here. Different note files exist for each collaborator: [Sarah](https://github.com/srmarzec/albopictus_remapping/blob/main/misc/SarahNotes.md), [Angela](https://github.com/srmarzec/albopictus_remapping/blob/main/misc/Angelanotes.md), and [Mackenzie](https://github.com/srmarzec/albopictus_remapping/blob/main/misc/Mackenzienotes.md).

Information about sequence analysis (raw read count, read alignment rate, etc.) for each dataset can be found [here](https://docs.google.com/spreadsheets/d/1hIqqMIk8ZVw56BJ8_YN_OnwuzdWBSH7bujwDMTqsCKs/edit?usp=sharing)

In a bid to be more efficient in requesting future resources for running jobs, a [list](https://docs.google.com/spreadsheets/d/1nUIWIHBaFrsp9vRli46SlE7K6ci-MSl6gKQy1SEJJw0/edit?usp=sharing) was generated for the larval dataset with both the requested and actually used resources. Note, that the scripts linked below use the the former, so this list serves as an example moving forward. 

## Data Retrieval
SRA files from NCBI for each set of samples were downloaded and converted to fastq.gz format. See [notes](https://github.com/srmarzec/albopictus_remapping/blob/main/misc/sra_accession/sraRetrievalTips.md) for details.

## Cleaning Data
Preliminary tests showed no need for preprocessing raw data beside trimmomatic steps (read alignment rate is ~82%).

### Trimmomatic
Trimmomatic version 0.39 [code for download](https://github.com/srmarzec/albopictus_remapping/blob/main/misc/SarahNotes.md#downloading-the-right-version-of-trimmomatic)

Generic script: [trim.sh](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/trim.sh)

We decided to use this adapter file because based on the manual, it was most likely used for HiSeq machines (which all sets of our data were run on). Note that all parameter for trimmomatic will run in the order they are written. We decided to headcrop (remove) the first 15 bases as they are 'messy' from "random" hexamer priming that occurs during RNA-seq library preparation.

### Quality Control

Fastqc [code for download](https://github.com/srmarzec/albopictus_remapping/blob/main/misc/SarahNotes.md#downloadingusing-fastqc)

Fastqc [script](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/fastqc.sh)

## Mapping
We will be mapping with STAR (v2.7.1a). 

Our reference genome is a new one generated by PI's group, but we have access through him.

With Star, the genome should be indexed first: [STAR_genomeIndex.sh](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/STAR_genomeIndex.sh)

Then mapped: [STAR_map.sh](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/STAR_map.sh)

Lastly, the sam files converted to bam and indexed for future work: [sam2bam.sh](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/sam2bam.sh)

## Counting Data

To count data, we will be using HTSeq

The general script is [htseq_count.sh](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/htseq_count.sh)

However, it may be that the stranded setting needs to be changed for the different datasets. Please check [here](https://github.com/srmarzec/albopictus_remapping/blob/main/misc/TestingStrandedness.md) for testing strandedness of data.

## Differential Expression with DESeq
After bringing htseq count files to local machine, we counted reads assigned to genes with this [script](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/Sarah/CountReadsAssignedToGenes.R).

Then, [DESeq](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/Sarah/DESeq.R) followed by [GO term enrichment](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/Sarah/GOenrichment.R) and [KEGG pathway id](https://github.com/srmarzec/albopictus_remapping/blob/main/scripts/Sarah/KEGG.R)
